packages/yew-router/src/hidden.rs

```diff
@@ -0,0 +1,407 @@
+use std::any::{Any, TypeId};
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::fmt::Display;
+use std::mem;
+use std::str::FromStr;
+
+use percent_encoding::{percent_decode_str, percent_encode, NON_ALPHANUMERIC};
+use regex::RegexSet;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use wasm_bindgen::JsValue;
+
+use crate::history::Route;
+pub use crate::Routable;
+
+/// This trait is implemented by the derive macro. There is a blanket implementation of
+/// `Routeable` for all `DerivedRoutable`.
+pub trait DerivedRoutable: PartialEq + Any + Default + Clone {
+    const VARIANTS: &'static [RouteMuncher<Self>];
+
+    fn to_args(&self) -> (usize, Args);
+}
+
+/// Lazily builds and caches a `RegexSet` keyed on the TypeId of the `DerivedRoutable`.
+fn cached_regex_set<T: DerivedRoutable>() -> &'static RegexSet {
+    thread_local! {
+        static REGEX_CACHE: RefCell<HashMap<TypeId, &'static RegexSet>> = RefCell::new(HashMap::new());
+    }
+
+    // In the fast path, we simply lookup the TypeId in the map.
+    REGEX_CACHE.with(|cache| {
+        *cache
+            .borrow_mut()
+            .entry(TypeId::of::<T>())
+            .or_insert_with(|| {
+                // In the slow path we build a `RegexSet` from all the individual variant regexes. Leaking
+                // this `RegexSet` is fine, because the total number of distinct `DerivedRoutable` types is
+                // fixed at compile time.
+                Box::leak(Box::new(
+                    RegexSet::new(T::VARIANTS.iter().map(|v| v.regex))
+                        .expect("Invalid routing regex"),
+                ))
+            })
+    })
+}
+
+/// Blanket implementation of `Routable` for `DerivedRoutable`.
+impl<T: DerivedRoutable> Routable for T {
+    fn from_route(route: &Route) -> Result<Self, Self>
+    where
+        Self: Sized,
+    {
+        // Use a `RegexSet` to rapidly narrow the search to a small number of
+        // candidate routes.
+        let regex_set = cached_regex_set::<T>();
+        for candidate in regex_set.matches(&route.path) {
+            // Try each route in turn. `RegexSet` guarantees iteration order to
+            // match the original array.
+            let variant = &T::VARIANTS[candidate];
+            if let Some(res) = variant.from_route(route) {
+                return Ok(res);
+            }
+        }
+        Err(T::default())
```

Comment:
Using `Default` requires that a redirect route must be provided. We should use separate function that optionally returns a route for this case so we don't force user to define a 404 page.

Reply:
The problem with this is that you then need to make all the functions that deal with a route return an `Option` or a`Result` instead, and you always have to consider that error case. With `#[bind(rest)]` there is an easier option:

```rust
enum MyRoute {
    #[at("/")]
    Home,
    #[at("/foo")]
    Foo,
    #[at("/bar")]
    Bar,
    #[at("/")]
    Unknown(#[bind(rest)] Route),
}

impl Default for MyRoute {
    fn default() -> Self {
        Self::Home
    }
}
```

Now the `Unknown` case will act as a catch-all, and you can handle that however you want without redirecting.

However, it would be nice to avoid needing a `Default` implementation for this case when you have an `Unknown` variant, so my plan was to extend the macro to handle this better at a later time.

Reply:
That sounds reasonable. I think this way, although this is different from basically every other router implementation that I've seen, it's more Rust-y as it forces the user to deal with the not found case.

Reply:
Yeah - I figure there's 3 cases:

1) You don't care about other routes or defining a 404 page. In this case you can just have `impl Default` return `Home` or w/e, and unknown pages will get redirected to that.

2) You want other routes to get redirected to a special 404 page. In this case you define your 404 route in your enum and handle it as you would any other route.

3) You want to show a 404 page but not redirect them. In this case you define an `Unknown` variant in your enum as shown. Ideally you would not need a `Default` implementation at all but that can't be avoided for now.

In no case do you want to be dealing with a `None` or `Err` case from the current route function.

If there are any other cases you can think of, then I will consider those too.

Reply:
I can't think of any other case. Let's leave this as-is for now. We can always update this should any other case come up.